src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java

```diff
@@ -0,0 +1,185 @@
+			Assertions.assertThat(info.getTags()).isEmpty();
+		}
+
+		@CartesianProductTest(factory = "supplyValues")
+		@DisplayName("Can have an explicit factory method over the implicit one")
+		void explicitFactory(String string, TimeUnit unit) {
+			Assertions.assertThat(string).isIn("War", "Peace");
+			Assertions.assertThat(unit.name()).endsWith("S");
+		}
+
+	}
+
+	static CartesianProductTest.Sets nFold() {
+		return new CartesianProductTest.Sets()
+				.add("Alpha", "Omega")
+				.add(Runnable.class, Comparable.class, TestInfo.class)
+				.add(TimeUnit.DAYS, TimeUnit.HOURS);
+	}
+
+	static CartesianProductTest.Sets supplyValues() {
+		return new CartesianProductTest.Sets().add("War", "Peace").add(TimeUnit.SECONDS, TimeUnit.DAYS);
+	}
+
+	@Nested
+	class BadConfigurationTests {
+
+		@Test
+		@DisplayName("Test fails when there is no factory method")
+		void throwsForMissingFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "noFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("not found");
+		}
+
+		@Test
+		@DisplayName("Test fails when the factory is not static")
+		void throwsForNonStaticFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "nonStaticFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("must be static");
+		}
+
+		@Test
+		@DisplayName("Test fails when the factory does not return Sets")
+		void throwsForWrongReturnValueFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "wrongReturnFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("must return");
+		}
+
+		@Test
+		@DisplayName("Test fails if the factory does not produce enough parameters")
+		void throwsForTooFewFactoryMethodParameters() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "incompleteFactory", int.class,
+						String.class, TimeUnit.class);
+
+			assertThat(results).hasNumberOfDynamicallyRegisteredTests(3).hasNumberOfFailedTests(3);
+		}
+
+		@Test
+		@DisplayName("Test fails if the factory produces too much parameters")
+		void throwsForTooManyFactoryMethodParameters() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "bloatedFactory", int.class,
+						String.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("must register values for each parameter");
```

Comment:
Why is it possible to generate a message when there a too much parameters, but not too less (see test above)?

Reply:
Don't know how to test for too few parameters, unless we check for auto-injected parameters (like `TestReporter` or `TestInfo`).

Reply:
Okay.